[Bug] Rust: error while compiling: "openAnkiDroidCollection": Anki already open, or media currently syncing.

```
android.database.sqlite.SQLiteException: error while compiling: "openAnkiDroidCollection": Anki already open, or media currently syncing.
at net.ankiweb.rsdroid.BackendV1Impl.openAnkiDroidCollection(BackendV1Impl.java:10)
at net.ankiweb.rsdroid.BackendMutex.openAnkiDroidCollection(BackendMutex.java:2)
at net.ankiweb.rsdroid.BackendUtils.openAnkiDroidCollection(BackendUtils.java:1)
at net.ankiweb.rsdroid.database.RustSupportSQLiteOpenHelper.createRustSupportSQLiteDatabase(RustSupportSQLiteOpenHelper.java:4)
at net.ankiweb.rsdroid.database.RustSupportSQLiteOpenHelper.getWritableDatabase(RustSupportSQLiteOpenHelper.java:2)
at com.ichi2.libanki.DB.<init>(DB.java:9)
at com.ichi2.libanki.backend.RustDroidBackend.openCollectionDatabase(RustDroidBackend.java:1)
at com.ichi2.libanki.Storage.Collection(Storage.java:6)
at com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:8)
at com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:3)
at com.ichi2.anki.CollectionHelper.getColSafe(CollectionHelper.java:1)
at com.ichi2.anki.DeckPicker.firstCollectionOpen(DeckPicker.java:11)
at com.ichi2.anki.DeckPicker.onCreate(DeckPicker.java:5)
at android.app.Activity.performCreate(Activity.java:7963)
at android.app.Activity.performCreate(Activity.java:7952)
at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1307)
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3629)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3806)
at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2267)
at android.os.Handler.dispatchMessage(Handler.java:107)
at android.os.Looper.loop(Looper.java:237)
at android.app.ActivityThread.main(ActivityThread.java:8167)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
Caused by: net.ankiweb.rsdroid.BackendException$BackendDbException: Anki already open, or media currently syncing.
at net.ankiweb.rsdroid.RustBackendImpl.validateMessage(RustBackendImpl.java:3)
at net.ankiweb.rsdroid.BackendV1Impl.openAnkiDroidCollection(BackendV1Impl.java:6)
... 27 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
@@ -193,6 +193,9 @@ public class AnkiDroidApp extends Application {
     @Nullable
     private Throwable mWebViewError;
 
+    /** HACK: Whether an exception report has been thrown - TODO: Rewrite an ACRA Listener to do this */
+    @VisibleForTesting
+    public static boolean sSentExceptionReportHack;
 
     @NonNull
     public static InputStream getResourceAsStream(@NonNull String name) {
@@ -408,6 +411,8 @@ public class AnkiDroidApp extends Application {
 
     public static void sendExceptionReport(Throwable e, String origin, @Nullable String additionalInfo, boolean onlyIfSilent) {
         UsageAnalytics.sendAnalyticsException(e, false);
+
+        sSentExceptionReportHack = true;
 
         if (onlyIfSilent) {
             String reportMode = getSharedPrefs(getInstance().getApplicationContext()).getString(AnkiDroidApp.FEEDBACK_REPORT_KEY, FEEDBACK_REPORT_ASK);

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
@@ -33,6 +33,8 @@ import com.ichi2.libanki.utils.Time;
 import com.ichi2.preferences.PreferenceExtensions;
 import com.ichi2.utils.FileUtil;
 
+import net.ankiweb.rsdroid.BackendException;
+
 import java.io.File;
 import java.io.IOException;
 
@@ -151,6 +153,9 @@ public class CollectionHelper {
     public synchronized Collection getColSafe(Context context) {
         try {
             return getCol(context);
+        } catch (BackendException.BackendDbException.BackendDbLockedException e) {
+            Timber.w(e);
+            return null;
         } catch (Exception e) {
             Timber.w(e);
             AnkiDroidApp.sendExceptionReport(e, "CollectionHelper.getColSafe");

--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
@@ -173,7 +173,8 @@ public class DeckPicker extends NavigationDrawerActivity implements
     /**
      * Available options performed by other activities (request codes for onActivityResult())
      */
-    private static final int REQUEST_STORAGE_PERMISSION = 0;
+    @VisibleForTesting
+    static final int REQUEST_STORAGE_PERMISSION = 0;
     private static final int REQUEST_PATH_UPDATE = 1;
     public static final int REPORT_FEEDBACK = 4;
     private static final int LOG_IN_FOR_SYNC = 6;
@@ -524,20 +525,30 @@ public class DeckPicker extends NavigationDrawerActivity implements
         mReviewSummaryTextView = findViewById(R.id.today_stats_text_view);
 
         Timber.i("colOpen: %b", colOpen);
-        if (colOpen) {
-            // Show any necessary dialogs (e.g. changelog, special messages, etc)
-            showStartupScreensAndDialogs(preferences, 0);
-        } else {
-            // Show error dialogs
-            if (Permissions.hasStorageAccessPermission(this)) {
-                StartupFailure failure = InitialActivity.getStartupFailureType(this);
-                handleStartupFailure(failure);
-            }
-            // firstCollectionOpen should have been called to show a dialog - so no need to do anything here.
+        // if permission is denied, firstCollectionOpen() requests it and onRequestPermissionsResult continues execution
+        if (Permissions.hasStorageAccessPermission(this)) {
+            handleStartup(colOpen);
         }
+
 
         mShortAnimDuration = getResources().getInteger(android.R.integer.config_shortAnimTime);
     }
+
+    /** The first call in showing dialogs for startup - error or success */
+    private void handleStartup(boolean colOpen) {
+        // TODO: colOpen is not colIsOpen() if called from onCreate - we should fix this mismatch of terms
+        // or use the same variable if the semantics should have been equivalent
+        if (colOpen) {
+            // Show any necessary dialogs (e.g. changelog, special messages, etc)
+            SharedPreferences sharedPrefs = AnkiDroidApp.getSharedPrefs(this);
+            showStartupScreensAndDialogs(sharedPrefs, 0);
+        } else {
+            // Show error dialogs
+            StartupFailure failure = InitialActivity.getStartupFailureType(this);
+            handleStartupFailure(failure);
+        }
+    }
+
 
     @VisibleForTesting
     void handleStartupFailure(StartupFailure failure) {
@@ -926,7 +937,7 @@ public class DeckPicker extends NavigationDrawerActivity implements
         if (requestCode == REQUEST_STORAGE_PERMISSION && permissions.length == 1) {
             if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                 invalidateOptionsMenu();
-                showStartupScreensAndDialogs(AnkiDroidApp.getSharedPrefs(this), 0);
+                handleStartup(colIsOpen());
             } else {
                 // User denied access to file storage  so show error toast and display "App Info"
                 Toast.makeText(this, R.string.startup_no_storage_permission, Toast.LENGTH_LONG).show();

--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
@@ -22,7 +22,6 @@
 package com.ichi2.anki;
 
 import android.Manifest;
-import android.animation.Animator;
 import android.content.BroadcastReceiver;
 import android.content.Context;
 import android.content.DialogInterface;
@@ -44,7 +43,6 @@ import android.os.ParcelFileDescriptor;
 import android.provider.Settings;
 
 import com.afollestad.materialdialogs.GravityEnum;
-import com.google.android.material.floatingactionbutton.FloatingActionButton;
 import com.google.android.material.snackbar.Snackbar;
 
 import androidx.annotation.NonNull;
@@ -86,6 +84,7 @@ import android.widget.Toast;
 
 import com.afollestad.materialdialogs.MaterialDialog;
 import com.ichi2.anki.CollectionHelper.CollectionIntegrityStorageCheck;
+import com.ichi2.anki.InitialActivity.StartupFailure;
 import com.ichi2.anki.StudyOptionsFragment.StudyOptionsListener;
 import com.ichi2.anki.analytics.UsageAnalytics;
 import com.ichi2.anki.dialogs.AsyncDialogFragment;
@@ -531,34 +530,36 @@ public class DeckPicker extends NavigationDrawerActivity implements
         } else {
             // Show error dialogs
             if (Permissions.hasStorageAccessPermission(this)) {
-                if (!AnkiDroidApp.isSdCardMounted()) {
-                    Timber.i("SD card not mounted");
-                    onSdCardNotMounted();
-                } else if (!CollectionHelper.isCurrentAnkiDroidDirAccessible(this)) {
-                    Timber.i("AnkiDroid directory inaccessible");
-                    Intent i = Preferences.getPreferenceSubscreenIntent(this, "com.ichi2.anki.prefs.advanced");
-                    startActivityForResultWithoutAnimation(i, REQUEST_PATH_UPDATE);
-                    Toast.makeText(this, R.string.directory_inaccessible, Toast.LENGTH_LONG).show();
-                } else if (isFutureAnkiDroidVersion()) {
-                    Timber.i("Displaying database versioning");
-                    showDatabaseErrorDialog(DatabaseErrorDialog.INCOMPATIBLE_DB_VERSION);
-                } else {
-                    Timber.i("Displaying database error");
-                    showDatabaseErrorDialog(DatabaseErrorDialog.DIALOG_LOAD_FAILED);
-                }
+                StartupFailure failure = InitialActivity.getStartupFailureType(this);
+                handleStartupFailure(failure);
             }
+            // firstCollectionOpen should have been called to show a dialog - so no need to do anything here.
         }
 
         mShortAnimDuration = getResources().getInteger(android.R.integer.config_shortAnimTime);
     }
 
-
-    private boolean isFutureAnkiDroidVersion() {
-        try {
-            return CollectionHelper.isFutureAnkiDroidVersion(this);
-        } catch (Exception e) {
-            Timber.w(e, "Could not determine if future AnkiDroid version - assuming not");
-            return false;
+    @VisibleForTesting
+    void handleStartupFailure(StartupFailure failure) {
+        switch (failure) {
+            case SD_CARD_NOT_MOUNTED:
+                Timber.i("SD card not mounted");
+                onSdCardNotMounted();
+                break;
+            case DIRECTORY_NOT_ACCESSIBLE:
+                Timber.i("AnkiDroid directory inaccessible");
+                Intent i = Preferences.getPreferenceSubscreenIntent(this, "com.ichi2.anki.prefs.advanced");
+                startActivityForResultWithoutAnimation(i, REQUEST_PATH_UPDATE);
+                Toast.makeText(this, R.string.directory_inaccessible, Toast.LENGTH_LONG).show();
+                break;
+            case FUTURE_ANKIDROID_VERSION:
+                Timber.i("Displaying database versioning");
+                showDatabaseErrorDialog(DatabaseErrorDialog.INCOMPATIBLE_DB_VERSION);
+                break;
+            case DB_ERROR:
+            default:
+                Timber.i("Displaying database error");
+                showDatabaseErrorDialog(DatabaseErrorDialog.DIALOG_LOAD_FAILED);
         }
     }
 

--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
@@ -556,6 +556,10 @@ public class DeckPicker extends NavigationDrawerActivity implements
                 Timber.i("Displaying database versioning");
                 showDatabaseErrorDialog(DatabaseErrorDialog.INCOMPATIBLE_DB_VERSION);
                 break;
+            case DATABASE_LOCKED:
+                Timber.i("Displaying database locked error");
+                showDatabaseErrorDialog(DatabaseErrorDialog.DIALOG_DB_LOCKED);
+                break;
             case DB_ERROR:
             default:
                 Timber.i("Displaying database error");

--- a/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.java
@@ -1,0 +1,61 @@
+/*
+ *  Copyright (c) 2021 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki;
+
+import android.content.Context;
+
+import androidx.annotation.CheckResult;
+import androidx.annotation.NonNull;
+import timber.log.Timber;
+
+/** Utilities for launching the first activity (currently the DeckPicker) */
+public class InitialActivity {
+
+    private InitialActivity() {
+
+    }
+
+    @NonNull
+    @CheckResult
+    public static StartupFailure getStartupFailureType(Context context) {
+        if (!AnkiDroidApp.isSdCardMounted()) {
+            return StartupFailure.SD_CARD_NOT_MOUNTED;
+        } else if (!CollectionHelper.isCurrentAnkiDroidDirAccessible(context)) {
+            return StartupFailure.DIRECTORY_NOT_ACCESSIBLE;
+        } else if (isFutureAnkiDroidVersion(context)) {
+            return StartupFailure.FUTURE_ANKIDROID_VERSION;
+        } else {
+            return StartupFailure.DB_ERROR;
+        }
+    }
+
+    private static boolean isFutureAnkiDroidVersion(Context context) {
+        try {
+            return CollectionHelper.isFutureAnkiDroidVersion(context);
+        } catch (Exception e) {
+            Timber.w(e, "Could not determine if future AnkiDroid version - assuming not");
+            return false;
+        }
+    }
+
+    public enum StartupFailure {
+        SD_CARD_NOT_MOUNTED,
+        DIRECTORY_NOT_ACCESSIBLE,
+        FUTURE_ANKIDROID_VERSION,
+        DB_ERROR,
+    }
+}

--- a/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.java
@@ -18,6 +18,8 @@ package com.ichi2.anki;
 
 import android.content.Context;
 
+import net.ankiweb.rsdroid.BackendException;
+
 import androidx.annotation.CheckResult;
 import androidx.annotation.NonNull;
 import timber.log.Timber;
@@ -39,6 +41,13 @@ public class InitialActivity {
         } else if (isFutureAnkiDroidVersion(context)) {
             return StartupFailure.FUTURE_ANKIDROID_VERSION;
         } else {
+            try {
+                CollectionHelper.getInstance().getCol(context);
+            } catch (BackendException.BackendDbException.BackendDbLockedException e) {
+                return StartupFailure.DATABASE_LOCKED;
+            } catch (Exception ignored) {
+
+            }
             return StartupFailure.DB_ERROR;
         }
     }
@@ -57,5 +66,6 @@ public class InitialActivity {
         DIRECTORY_NOT_ACCESSIBLE,
         FUTURE_ANKIDROID_VERSION,
         DB_ERROR,
+        DATABASE_LOCKED,
     }
 }

--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java
@@ -41,7 +41,7 @@ public class DatabaseErrorDialog extends AsyncDialogFragment {
     public static final int DIALOG_FULL_SYNC_FROM_SERVER = 8;
     /** If the database is locked, all we can do is reset the app */
     public static final int DIALOG_DB_LOCKED = 9;
-    /** If the datbase is at a version higher than what we can currently handle */
+    /** If the database is at a version higher than what we can currently handle */
     public static final int INCOMPATIBLE_DB_VERSION = 10;
 
     // public flag which lets us distinguish between inaccessible and corrupt database

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/backend/DroidBackendFactory.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/backend/DroidBackendFactory.java
@@ -23,11 +23,16 @@ import net.ankiweb.rsdroid.BackendFactory;
 import net.ankiweb.rsdroid.RustBackendFailedException;
 import net.ankiweb.rsdroid.RustCleanup;
 
+import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
 import timber.log.Timber;
 
 /** Responsible for selection of either the Rust or Java-based backend */
 public class DroidBackendFactory {
+
+    private static DroidBackend sBackendForTesting;
+
 
     /** Intentionally private - use {@link DroidBackendFactory#getInstance(boolean)}} */
     private DroidBackendFactory() {
@@ -38,8 +43,12 @@ public class DroidBackendFactory {
      * Obtains an instance of a {@link DroidBackend}.
      * Each call to this method will generate a separate instance which can handle a new Anki collection
      */
+    @NonNull
     @RustCleanup("Change back to a constant SYNC_VER")
     public static DroidBackend getInstance(boolean useBackend) {
+        if (sBackendForTesting != null) {
+            return sBackendForTesting;
+        }
 
         BackendFactory backendFactory = null;
         if (useBackend) {
@@ -63,5 +72,10 @@ public class DroidBackendFactory {
         } else {
             return new RustDroidBackend(backendFactory);
         }
+    }
+
+    @VisibleForTesting
+    public static void setOverride(DroidBackend backend) {
+        sBackendForTesting = backend;
     }
 }

--- a/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
@@ -4,6 +4,7 @@ import android.content.Context;
 import android.content.Intent;
 import android.content.SharedPreferences;
 import android.content.SharedPreferences.Editor;
+import android.content.pm.PackageManager;
 
 import com.ichi2.anki.dialogs.DatabaseErrorDialog;
 import com.ichi2.libanki.Collection;
@@ -221,12 +222,37 @@ public class DeckPickerTest extends RobolectricTest {
         }
     }
 
+    @Test
+    public void databaseLockedNoPermissionIntegrationTest() {
+        // no permissions -> grant permissions -> db locked
+        try {
+            InitialActivityTest.setupForDefault();
+            BackendEmulatingOpenConflict.enable();
+
+            DeckPickerEx d = super.startActivityNormallyOpenCollectionWithIntent(DeckPickerEx.class, new Intent());
+
+            // grant permissions
+            InitialActivityTest.setupForDatabaseConflict();
+
+            d.onStoragePermissionGranted();
+
+            assertThat("A specific dialog for a conflict should be shown", d.mDatabaseErrorDialog, is(DatabaseErrorDialog.DIALOG_DB_LOCKED));
+        } finally {
+            BackendEmulatingOpenConflict.disable();
+            InitialActivityTest.setupForDefault();
+        }
+    }
+
     private static class DeckPickerEx extends DeckPicker {
         private int mDatabaseErrorDialog;
 
         @Override
         public void showDatabaseErrorDialog(int id) {
             this.mDatabaseErrorDialog = id;
+        }
+
+        public void onStoragePermissionGranted() {
+            onRequestPermissionsResult(DeckPicker.REQUEST_STORAGE_PERMISSION, new String[] { "" }, new int[] { PackageManager.PERMISSION_GRANTED });
         }
     }
 }

--- a/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
@@ -209,6 +209,7 @@ public class DeckPickerTest extends RobolectricTest {
 
     @Test
     public void databaseLockedWithPermissionIntegrationTest() {
+        AnkiDroidApp.sSentExceptionReportHack = false;
         try {
             BackendEmulatingOpenConflict.enable();
             InitialActivityTest.setupForDatabaseConflict();
@@ -216,6 +217,8 @@ public class DeckPickerTest extends RobolectricTest {
             DeckPickerEx d = super.startActivityNormallyOpenCollectionWithIntent(DeckPickerEx.class, new Intent());
 
             assertThat("A specific dialog for a conflict should be shown", d.mDatabaseErrorDialog, is(DatabaseErrorDialog.DIALOG_DB_LOCKED));
+
+            assertThat("No exception reports should be thrown", AnkiDroidApp.sSentExceptionReportHack, is(false));
         } finally {
             BackendEmulatingOpenConflict.disable();
             InitialActivityTest.setupForDefault();

--- a/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
@@ -1,23 +1,25 @@
 package com.ichi2.anki;
 
 import android.content.Context;
-
-import androidx.test.core.app.ActivityScenario;
-import androidx.test.ext.junit.runners.AndroidJUnit4;
-
 import android.content.Intent;
 import android.content.SharedPreferences;
 import android.content.SharedPreferences.Editor;
 
+import com.ichi2.anki.dialogs.DatabaseErrorDialog;
 import com.ichi2.libanki.Collection;
 import com.ichi2.libanki.DeckConfig;
 import com.ichi2.libanki.sched.AbstractSched;
+import com.ichi2.testutils.BackendEmulatingOpenConflict;
 
 import org.junit.Test;
 import org.junit.runner.RunWith;
+import org.robolectric.Robolectric;
 
 import java.util.HashMap;
 import java.util.Map;
+
+import androidx.test.core.app.ActivityScenario;
+import androidx.test.ext.junit.runners.AndroidJUnit4;
 
 import static com.ichi2.anki.DeckPicker.UPGRADE_VERSION_KEY;
 import static org.hamcrest.MatcherAssert.assertThat;
@@ -192,5 +194,39 @@ public class DeckPickerTest extends RobolectricTest {
         advanceRobolectricLooperWithSleep();
 
         assertThat("deck was deleted", getCol().getDecks().count(), is(1));
+    }
+
+    @Test
+    public void databaseLockedTest() {
+        // don't call .onCreate
+        DeckPickerEx deckPicker = Robolectric.buildActivity(DeckPickerEx.class, new Intent()).get();
+
+        deckPicker.handleStartupFailure(InitialActivity.StartupFailure.DATABASE_LOCKED);
+
+        assertThat(deckPicker.mDatabaseErrorDialog, is(DatabaseErrorDialog.DIALOG_DB_LOCKED));
+    }
+
+    @Test
+    public void databaseLockedWithPermissionIntegrationTest() {
+        try {
+            BackendEmulatingOpenConflict.enable();
+            InitialActivityTest.setupForDatabaseConflict();
+
+            DeckPickerEx d = super.startActivityNormallyOpenCollectionWithIntent(DeckPickerEx.class, new Intent());
+
+            assertThat("A specific dialog for a conflict should be shown", d.mDatabaseErrorDialog, is(DatabaseErrorDialog.DIALOG_DB_LOCKED));
+        } finally {
+            BackendEmulatingOpenConflict.disable();
+            InitialActivityTest.setupForDefault();
+        }
+    }
+
+    private static class DeckPickerEx extends DeckPicker {
+        private int mDatabaseErrorDialog;
+
+        @Override
+        public void showDatabaseErrorDialog(int id) {
+            this.mDatabaseErrorDialog = id;
+        }
     }
 }

--- a/AnkiDroid/src/test/java/com/ichi2/anki/InitialActivityTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/InitialActivityTest.java
@@ -1,0 +1,70 @@
+/*
+ *  Copyright (c) 2021 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki;
+
+import android.app.Application;
+
+import com.ichi2.testutils.BackendEmulatingOpenConflict;
+import com.ichi2.testutils.EmptyApplication;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.Shadows;
+import org.robolectric.annotation.Config;
+import org.robolectric.shadows.ShadowApplication;
+import org.robolectric.shadows.ShadowEnvironment;
+
+import androidx.test.core.app.ApplicationProvider;
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@RunWith(AndroidJUnit4.class)
+@Config(application = EmptyApplication.class) // no point in Application init if we don't use it
+public class InitialActivityTest extends RobolectricTest {
+    @Before
+    @Override
+    public void setUp() {
+        super.setUp();
+        BackendEmulatingOpenConflict.enable();
+    }
+
+    @After
+    @Override
+    public void tearDown() {
+        super.tearDown();
+        BackendEmulatingOpenConflict.disable();
+    }
+
+    @Test
+    public void testInitialActivityResult() {
+        setupForDatabaseConflict();
+
+        InitialActivity.StartupFailure f = InitialActivity.getStartupFailureType(getTargetContext());
+
+        assertThat("A conflict should be returned", f, is(InitialActivity.StartupFailure.DATABASE_LOCKED));
+    }
+
+    public static void setupForDatabaseConflict() {
+        ShadowApplication app = Shadows.shadowOf((Application) ApplicationProvider.getApplicationContext());
+        app.grantPermissions(android.Manifest.permission.WRITE_EXTERNAL_STORAGE, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+        ShadowEnvironment.setExternalStorageState("mounted");
+    }
+}

--- a/AnkiDroid/src/test/java/com/ichi2/anki/InitialActivityTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/InitialActivityTest.java
@@ -55,16 +55,26 @@ public class InitialActivityTest extends RobolectricTest {
 
     @Test
     public void testInitialActivityResult() {
-        setupForDatabaseConflict();
+        try {
+            setupForDatabaseConflict();
 
-        InitialActivity.StartupFailure f = InitialActivity.getStartupFailureType(getTargetContext());
+            InitialActivity.StartupFailure f = InitialActivity.getStartupFailureType(getTargetContext());
 
-        assertThat("A conflict should be returned", f, is(InitialActivity.StartupFailure.DATABASE_LOCKED));
+            assertThat("A conflict should be returned", f, is(InitialActivity.StartupFailure.DATABASE_LOCKED));
+        } finally {
+            setupForDefault();
+        }
     }
 
     public static void setupForDatabaseConflict() {
         ShadowApplication app = Shadows.shadowOf((Application) ApplicationProvider.getApplicationContext());
         app.grantPermissions(android.Manifest.permission.WRITE_EXTERNAL_STORAGE, android.Manifest.permission.READ_EXTERNAL_STORAGE);
         ShadowEnvironment.setExternalStorageState("mounted");
+    }
+
+    public static void setupForDefault() {
+        ShadowApplication app = Shadows.shadowOf((Application) ApplicationProvider.getApplicationContext());
+        app.denyPermissions(android.Manifest.permission.WRITE_EXTERNAL_STORAGE, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+        ShadowEnvironment.setExternalStorageState("removed");
     }
 }

--- a/AnkiDroid/src/test/java/com/ichi2/testutils/BackendEmulatingOpenConflict.java
+++ b/AnkiDroid/src/test/java/com/ichi2/testutils/BackendEmulatingOpenConflict.java
@@ -1,0 +1,60 @@
+/*
+ *  Copyright (c) 2021 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.testutils;
+
+import com.ichi2.libanki.DB;
+import com.ichi2.libanki.backend.DroidBackendFactory;
+import com.ichi2.libanki.backend.RustDroidBackend;
+
+import net.ankiweb.rsdroid.BackendException;
+import net.ankiweb.rsdroid.BackendFactory;
+import net.ankiweb.rsdroid.RustBackendFailedException;
+
+import BackendProto.Backend;
+
+import static org.mockito.Mockito.mock;
+
+/** Test helper:
+ * causes getCol to emulate an exception caused by having another AnkiDroid instance open on the same collection
+ */
+public class BackendEmulatingOpenConflict extends RustDroidBackend {
+
+    public BackendEmulatingOpenConflict(BackendFactory mBackend) {
+        super(mBackend);
+    }
+
+
+    public static void enable() {
+        try {
+            DroidBackendFactory.setOverride(new BackendEmulatingOpenConflict(BackendFactory.createInstance()));
+        } catch (RustBackendFailedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+
+    public static void disable() {
+        DroidBackendFactory.setOverride(null);
+    }
+
+
+    @Override
+    public DB openCollectionDatabase(String path) {
+        Backend.BackendError error = mock(Backend.BackendError.class);
+        throw new BackendException.BackendDbException.BackendDbLockedException(error);
+    }
+}

--- a/AnkiDroid/src/test/java/com/ichi2/testutils/BackendEmulatingOpenConflictTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/testutils/BackendEmulatingOpenConflictTest.java
@@ -1,0 +1,55 @@
+/*
+ *  Copyright (c) 2021 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.testutils;
+
+import com.ichi2.anki.CollectionHelper;
+import com.ichi2.anki.RobolectricTest;
+
+import net.ankiweb.rsdroid.BackendException.BackendDbException.BackendDbLockedException;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+
+import static org.junit.Assert.assertThrows;
+
+@RunWith(AndroidJUnit4.class)
+public class BackendEmulatingOpenConflictTest extends RobolectricTest {
+
+    @Before
+    @Override
+    public void setUp() {
+        super.setUp();
+        BackendEmulatingOpenConflict.enable();
+    }
+
+    @After
+    @Override
+    public void tearDown() {
+        super.tearDown();
+        BackendEmulatingOpenConflict.disable();
+    }
+
+    @Test
+    public void assumeMocksAreValid() {
+        assertThrows(BackendDbLockedException.class, () -> CollectionHelper.getInstance().getCol(super.getTargetContext()));
+    }
+
+}

